src/Avalonia.Controls/DockPanel.cs

```diff
@@ -75,39 +93,61 @@ public bool LastChildFill
             set => SetValue(LastChildFillProperty, value);
         }
 
+        /// <summary>
+        /// Gets or sets the horizontal distance between the child objects.
+        /// </summary>
+        public double HorizontalSpacing
+        {
+            get => GetValue(HorizontalSpacingProperty);
+            set => SetValue(HorizontalSpacingProperty, value);
+        }
+
+        /// <summary>
+        /// Gets or sets the vertical distance between the child objects.
+        /// </summary>
+        public double VerticalSpacing
+        {
+            get => GetValue(VerticalSpacingProperty);
+            set => SetValue(VerticalSpacingProperty, value);
+        }
+
+
         /// <summary>
         /// Updates DesiredSize of the DockPanel.  Called by parent Control.  This is the first pass of layout.
         /// </summary>
         /// <remarks>
         /// Children are measured based on their sizing properties and <see cref="Dock" />.  
-        /// Each child is allowed to consume all of the space on the side on which it is docked; Left/Right docked
+        /// Each child is allowed to consume all the space on the side on which it is docked; Left/Right docked
         /// children are granted all vertical space for their entire width, and Top/Bottom docked children are
         /// granted all horizontal space for their entire height.
         /// </remarks>
-        /// <param name="constraint">Constraint size is an "upper limit" that the return value should not exceed.</param>
+        /// <param name="availableSize">Constraint size is an "upper limit" that the return value should not exceed.</param>
         /// <returns>The Panel's desired size.</returns>
-        protected override Size MeasureOverride(Size constraint)
+        protected override Size MeasureOverride(Size availableSize)
         {
-            var children = Children;
+            var parentWidth = 0d;
+            var parentHeight = 0d;
+            var accumulatedWidth = 0d;
+            var accumulatedHeight = 0d;
+
+            var horizontalSpacing = false;
+            var verticalSpacing = false;
 
-            double parentWidth = 0;   // Our current required width due to children thus far.
-            double parentHeight = 0;   // Our current required height due to children thus far.
-            double accumulatedWidth = 0;   // Total width consumed by children.
-            double accumulatedHeight = 0;   // Total height consumed by children.
+            var visibleChildren = Children.Where(t => t.IsVisible).ToArray();
```

Comment:
Visible children should still be measured. This has the effect of immediately setting their `DesiredSize` to (0,0). Plus, this change also means that `LastChildFill` now applies to different controls depending on visibility, which would be a behavioral breaking change, and is generally unexpected.

You can revert to your previous code, simply don't add the spacing if the child is invisible.